python/cuml/neighbors/nearest_neighbors.pyx

```diff
@@ -393,5 +403,145 @@ class NearestNeighbors(Base):
+            distances = cp.ones(n_samples * n_neighbors)
+
+        elif mode == 'distance':
+            distances, indices = self.kneighbors(X, n_neighbors,
+                                                 return_cupy=True)
+            distances = cp.ravel(distances)
+
+        else:
+            raise ValueError('Unsupported mode, must be one of "connectivity" '
+                             'or "distance" but got "%s" instead' % mode)
+
+        n_samples = indices.shape[0]
+        n_samples_fit = self.X_m.shape[0]
+        n_nonzero = n_samples * n_neighbors
+        rowptr = cp.arange(0, n_nonzero + 1, n_neighbors)
+        return cp.sparse.csr_matrix((distances, cp.ravel(indices), rowptr),
+                                    shape=(n_samples, n_samples_fit))
+
+
+def kneighbors_graph(X=None, n_neighbors=5, mode='connectivity', verbose=False,
+                     handle=None, algorithm="brute", metric="minkowski", p=2,
+                     include_self=False, metric_params=None):
+    """
+    Computes the (weighted) graph of k-Neighbors for points in X.
+
+    Parameters
+    ----------
+    X : array-like (device or host) shape = (n_samples, n_features)
+        Dense matrix (floats or doubles) of shape (n_samples, n_features).
+        Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+        ndarray, cuda array interface compliant array like CuPy
+
+    n_neighbors : Integer
+        Number of neighbors to search. If not provided, the n_neighbors
+        from the model instance is used (default=5)
+
+    mode : string (default='connectivity')
+        Values in connectivity matrix: 'connectivity' returns the
+        connectivity matrix with ones and zeros, 'distance' returns the
+        edges as the Euclidean distance between points
+
+    verbose : int or boolean (default = False)
+        Logging level
+
+    handle : cumlHandle
+        The cumlHandle resources to use
+
+    algorithm : string (default='brute')
+        The query algorithm to use. Currently, only 'brute' is supported.
+
+    metric : string (default='minkowski').
+        Distance metric to use. Supported distances are ['l1, 'cityblock',
+        'taxicab', 'manhattan', 'euclidean', 'l2', 'braycurtis', 'canberra',
+        'minkowski', 'chebyshev', 'jensenshannon', 'cosine', 'correlation']
+
+    p : float (default=2) Parameter for the Minkowski metric. When p = 1, this
+        is equivalent to manhattan distance (l1), and euclidean distance (l2)
+        for p = 2. For arbitrary p, minkowski distance (lp) is used.
+
+    include_self : bool or 'auto' (default=False)
+        Whether or not to mark each sample as the first nearest neighbor to
+        itself. If 'auto', then True is used for mode='connectivity' and False
+        for mode='distance'.
+
+    metric_params : dict, optional (default = None) This is currently ignored.
+
+    Returns
+    -------
+    A: sparse graph in CSR format, shape = (n_samples, n_samples_fit)
```

Comment:
We will want to support the `output_type` argument in this function. It can be passed to the nearest neighbors instance underneath. We should also be a little more explicit about the types of supported outputs (e.g. they can be on host/device or ndarray/dataframe.)

Reply:
Added being more explicit about outputs but am not sure what the `output_type` argument provides, there's probably a gap in my knowledge. From my understanding, the input/output of cupy's CSR matrix is always cupy ndarrays. In that case, regardless of what the user chooses as the output_type, the kneighbors_graph function will always return a cupy CSR made of cupy.ndarrays?

Reply:
We should be supporting outputs on both host (Scipy CSR by calling cupy_csr.get()) and device. Additionally, we should support dataframes output in 3-column COO format. This should be on host or device as well. cuML diverges from Scikit-learn slightly in our ability to output pandas/cudf, but it should be simple enough by calling `cupy_csr.tocoo()` and converting the 3 underlying arrays to a `cudf/pandas.Series`

Reply:
Sounds good, I'll drop my code snippet of the DataFrame support in a comment below to save time should we want to add this in the future. An implementation that could be worth looking into is having cudf add some sparse support similar to what pandas has done (https://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html)